### ISSUE TITLE
Style/HashEachMethods is marked as a violation for non-hash objects

### ISSUE THREAD

**Issue by Fire-Dragon-DoL** [rubocop-hq/rubocop#8990]
## Expected behavior

Rubocop should mark violations only for Hash objects.

## Actual behavior

Rubocop marks any object with a `keys` method and a call to `each` on the keys a violation of this rule.
This cop breaks for the standard library `Thread.current.keys.each`, which does not provide a `each_key`.

## Steps to reproduce the problem

Create a file with:

```ruby
Thread.current.keys.each { |k| puts k.to_s }
```

Running rubocop or `rubocop -A` will report violations for this line.
The same is valid for any object with `keys` and `each` called on it:

```ruby
class Vault
  def self.keys
    []
  end
end
```

This type of code will also trigger the violation without having an actual `each_key` method.

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler. Here's an example:

```
$ rubocop -V
1.0.0 (using Parser 2.7.2.0, rubocop-ast 1.0.1, running on ruby 2.6.6 x86_64-darwin19)
```

**Comment by marcandre** [comment]
Please note that this cop is marked as unsafe for this reason. `rubocop -a` will not make an autocorrection, and `rubocop --safe` will not raise an offense at all. You can probably exclude all unsafe cops by default too.

As you probably realize, there is nothing that can be done for your `Vault` example (at least until RuboCop connects to [rbs](https://github.com/ruby/rbs)).

I guess we could exclude `Thread.current.keys` though, but any variation would still be flagged (`cur = Thread.current; cur.keys.each`). I'm not sure it's used enough to warrant it personally, but I would accept a PR that makes this extra check.

**Comment by Fire-Dragon-DoL** [comment]
Thanks @marcandre 

I won't provide a PR for this, from my perspective this rule is too prone to false positives and as such detrimental.
However, the input on `--safe` it's a good one, I'll discuss with some team members to run only those rules, since various others cause problems.

Even with `rbs`, not everything will be typed and as such this rule will keep having problems.
